src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/CachedExpressionCompiler.cs

```diff
@@ -38,117 +38,109 @@ private static class Compiler<TModel, TResult>
+                        return CompileCapturedConstant(memberExpression, constantExpression);
 
-                var constantExpression = expression.Body as ConstantExpression;
-                if (constantExpression != null)
-                {
-                    // model => {const}
+                    // model => StaticMember
+                    case MemberExpression memberExpression when memberExpression.Expression == null:
+                        return CompileFromStaticMemberAccess(expression, memberExpression);
+
+                    // model => model.Member
+                    case MemberExpression memberExpression when memberExpression.Expression == expression.Parameters[0]:
+                        return CompileFromMemberAccess(expression, memberExpression);
 
-                    var constantValue = (TResult)constantExpression.Value;
-                    return _ => constantValue;
+                    default:
+                        return CompileSlow(expression);
                 }
+            }
 
-                return null;
+            private static Func<TModel, TResult> CompileFromConstLookup(
+                ConstantExpression constantExpression)
+            {
+                // model => {const}
+                var constantValue = (TResult)constantExpression.Value;
+                return _ => constantValue;
             }
 
             private static Func<TModel, TResult> CompileFromIdentityFunc(
                 Expression<Func<TModel, TResult>> expression)
             {
-                if (expression == null)
-                {
-                    throw new ArgumentNullException(nameof(expression));
-                }
+                // model => model
 
-                if (expression.Body == expression.Parameters[0])
+                // Don't need to lock, as all identity funcs are identical.
+                if (_identityFunc == null)
                 {
-                    // model => model
-
-                    // Don't need to lock, as all identity funcs are identical.
-                    if (_identityFunc == null)
-                    {
-                        _identityFunc = expression.Compile();
-                    }
-
-                    return _identityFunc;
+                    _identityFunc = expression.Compile();
                 }
 
-                return null;
+                return _identityFunc;
             }
 
             private static Func<TModel, TResult> CompileFromMemberAccess(
-                Expression<Func<TModel, TResult>> expression)
+                Expression<Func<TModel, TResult>> expression,
+                MemberExpression memberExpression)
             {
-                if (expression == null)
+                // model => model.Member
+                if (_simpleMemberAccessCache.TryGetValue(memberExpression.Member, out var result))
```

Comment:
Why add this short-circuit?

If this is a performance measure, have you verified the extra complexity matters? If yes, add a comment indicating the benefit e.g. avoids creating the `_ => expression.Compile()` delegate.

Reply:
Not sure what you mean by short-circuit here.